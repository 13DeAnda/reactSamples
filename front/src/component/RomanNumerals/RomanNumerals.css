.convertedValue{
	font-size: 20px;
  margin-top: 30px;
}

.convertButton{
  margin-top: 10px;
}

.RomanNumerals input{
  text-align: center;
}