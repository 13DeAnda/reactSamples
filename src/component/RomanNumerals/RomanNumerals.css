.RomanNumerals{
	margin: 30px;
}