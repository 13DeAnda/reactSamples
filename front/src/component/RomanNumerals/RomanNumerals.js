import React from 'react';
import { Grid, Row } from 'react-bootstrap';
import './RomanNumerals.css';
import _ from 'lodash';


export default class RomanNumerals extends React.Component {

  constructor(props) {
    super(props);
    this.onChange = this.onChange.bind(this);
    this.romanToNumber = this.romanToNumber.bind(this);
    this.numberToRoman = this.numberToRoman.bind(this);
    this.convert = this.convert.bind(this);

    this.state = {
      textToConvert: "",
      converted: "",
      romanNumbers: {
        M: {number: 1000, extractor: {roman: "C", number: 100}, isExtractor: false},
        D: {number: 500, extractor:{roman: "C", number: 100}, isExtractor: false},
        C: {number: 100, extractor:{roman: "X", number: 10}, isExtractor: true, grouped: "D"},
        L: {number: 50, extractor:{roman: "X", number: 10}, isExtractor: false},
        X: {number: 10, extractor:{roman: "I", number: 1}, isExtractor: true, grouped:"L"},
        V: {number: 5, extractor:{roman: "I", number: 1}, isExtractor: false},
        I: {number: 1, extractor: null, isExtractor: true, grouped: "V"}
      }
    };
  }

  render() {
    return (
      <Grid className="RomanNumerals container text-center" >
        <Row className='title'> Convert Roman Numbers </Row>
        <Row>
          <input type="text" value={this.state.textToConvert} placeholder= "ex: XV or ex: 15" onChange={this.onChange} className='textInput'/><br/>
          <button onClick={this.convert}
                  className='convertButton'> Convert Numer </button><br/>
          <Row className={this.state.message? 'convertedValue error' : 'convertedValue'} >
            {this.state.message? this.state.message :this.state.converted}
          </Row>
        </Row>
      </Grid>
    );
  }

  //logic
  romanToNumber (valueToConvert){
    valueToConvert = valueToConvert.toUpperCase();
    var finalConvertedNumber = 0;

    for(var i = 0; i < valueToConvert.length; i++){
      var letter = valueToConvert[i];
      var nextLetter = valueToConvert[i+1];

      var letterObj = this.state.romanNumbers[letter];
      if(!letterObj){
        this.setState({message: "Invalid Input", converted: null});
        return "Invalid Input";
        break;
      }

      var nextLetterObj = nextLetter? this.state.romanNumbers[nextLetter] : null;

      if(nextLetter &&
        letterObj.isExtractor &&
        nextLetterObj.extractor &&
        nextLetterObj.extractor.roman === letter){

          finalConvertedNumber += (nextLetterObj.number - letterObj.number);
          i++;
      }
      else{
        finalConvertedNumber += letterObj.number;
      }
    }
    return finalConvertedNumber;
  }

  numberToRoman (valueToConvert){
    var residue = valueToConvert;
    var romanNumber = "";
    _.forEach(this.state.romanNumbers, function(item, key){
      if(residue % item.number < residue){
        var repeated = Math.floor(residue/item.number);
        residue = residue % item.number;
        if(repeated < 4 || key === "M"){
          for(var i =0; i < repeated; i++){
            romanNumber += key;
          }
        }
        else{
          romanNumber += key + item.grouped;
        }

      }
    });
    return romanNumber;
  }

  convert (valueToConvert){
    var value = "";
    if(isNaN(this.state.textToConvert)){
      value = this.romanToNumber(this.state.textToConvert)
    }
    else{
      value = this.numberToRoman(this.state.textToConvert);
    }

    this.setState({converted: value});
    return value;
  }

  onChange (e){
    this.setState({textToConvert : e.target.value, message: null, converted: null});
  }

}
