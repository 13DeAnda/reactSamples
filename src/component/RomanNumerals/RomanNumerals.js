import React from 'react';
import './RomanNumerals.css';
import { Grid } from 'react-bootstrap';
class RomanNumerals extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      started: false,
    };
  }

  romanToNumber = (arg) =>{
    this.setState({started: true});
    var romanNumbers = [
      {roman: "C", number: 1000, extractor: null, isExtractor: false},
      {roman: "D", number: 500, extractor:{roman: "C", number: "100"}, isExtractor: false},
      {roman: "C", number: 100, extractor:{roman: "X", number: "10"}, isExtractor: true},
      {roman: "L", number: 50, extractor:{roman: "X", number: "10"}, isExtractor: false},
      {roman: "X", number: 10, extractor:{roman: "I", number: "1"}, isExtractor: true},
      {roman: "V", number: 5, extractor:{roman: "I", number: "1"}, isExtractor: false},
      {roman: "I", number: 1, extractor: null, isExtractor: true}
    ];

    var finalConvertedNumber = 0;
    var numberToConvert = parseInt(arg);
    

    for(var i = 0; i < arg.length; i++){
      var letter = arg[i];
      console.log("the arg", arg[i]);
      for(var j = 0; j < romanNumbers.length; j++){
        var romanObj = romanNumbers[j];
        //if it has a next number check if the next number matches the extractor
        console.log("thorugh", romanObj, arg[i+1])
        if(letter === romanObj.roman && 
           arg[i+1] && 
           romanObj.isExtractor && 
           romanObj.extractor && 
           romanObj.extractor === arg[i+1]){

          finalConvertedNumber +=  (arg[i+1].number - romanObj.number);
          i++;
          break;
        }
        // else if(letter === romanObj.roman){
        //   finalConvertedNumber +=romanObj.number;
        //   break;
        // }
      }
      
    }
    console.log("the final", finalConvertedNumber);
  }

  render() {
    return (
      <Grid className="container-fluid RomanNumerals">
        {!this.state.started? this.romanToNumber('IV') : null}
      </Grid>
    );
  }
}
export default RomanNumerals;
