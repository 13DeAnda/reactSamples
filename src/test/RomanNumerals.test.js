import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import RomanNumerals from '../component/RomanNumerals/RomanNumerals';

describe('RomanNumerals component', () => {
  it('should render snapshot', () => {
    const component = renderer.create(
      <RomanNumerals />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

});
