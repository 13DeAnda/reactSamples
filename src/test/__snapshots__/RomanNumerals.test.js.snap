// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RomanNumerals component should render snapshot 1`] = `
<div
  className="container-fluid RomanNumerals container"
>
  <div
    className="text-center row"
  >
    <input
      className="textInput"
      onChange={[Function]}
      type="text"
      value=""
    />
    <br />
    <button
      className="convertButton"
      onClick={[Function]}
    >
       Convert Numer 
    </button>
    <br />
    <div
      className="convertedValue text-center row"
    >
      
    </div>
  </div>
</div>
`;
